src/core/service.c

```diff
@@ -2864,7 +2863,7 @@ static void service_notify_cgroup_empty_event(Unit *u) {
                  * SIGCHLD for. */
 
         case SERVICE_START:
-                if (s->type == SERVICE_NOTIFY) {
+                if (s->type == SERVICE_NOTIFY && main_pid_good(s) <= 0) {
```

Comment:
Ready notifications need not come from the main pid. For example:
```
[Service]
Type=notify
NotifyAccess=all
RemainAfterExit=yes
ExecStart=/bin/bash -c '(sleep 3; /usr/bin/systemd-notify --ready) &'
```
Hence the check would be more complicated: `&& main_pid_good(s) <= 0 && (!RemainAfterExit || NotifyAccess == "main")`. However, this effectively only focuses on the main process. Do we not receive a SIGCHLD event when the main process goes away anyway? In that case this whole block would be pretty useless and a better check would be `&& not_migrated_cgroup`. All we want to do is to figure out that we will not succeed anyway, so that we can fail early instead of waiting for a timeout.

Reply:
Not following. This is in the SERVICE_START phase, i.e. where the service isn't up yet. If Type=notify is configured, then the notify message can either come from the main process or from any other process in the cgroup. If there's no main process anymore, and nothing in the cgroup either, we can fail the unit right-away: there's no way how we'd ever be able to get the READY=1 message we are supposed to get because Type=notify is set. (Note that it is not supported in systemd to migrate processes to exiting services or scopes. A process can only migrate into newly started units, but never to existing ones).

RemainAfterExit is irrelevant for this all, because we are talking of SERVICE_START here, and if we don't get the sd_notify() there, and know we can't get it we'll never enter SERVICE_RUNNING which would be precondition to ever enter SERVICE_EXITED (i.e. the state that is enabled through RemainAfterExit)

NotifyAccess= is also not relevant: it can't be `none`, as Type=notify implies it's not set to `none`. Which means it has to be either main, all, or exec, but for all three of these cases we made sure that we can't get the READY=1 anymore, because the cgroup is dead and the non-cgroup PIDs are gone too.

Or to say this differently: note that the whole function is called only when it has been verified that the cgroup is empty. And if we expect READY=1, but there are not processes around that could possibly send this to us (or at least none which would get passed the NotifyAccess= check, then we can fail the unit.

Reply:
anyway, long story short: i don't grok what you are trying to say. could you elaborate, please?

Reply:
The sample service I provided forks off a non-main process that is allowed to send READY=1. When using PAM, this forked off process ends up outside the cgroup.

Judging from the code, there is also another way to get a notify service running, and that is via a reload request, which starts a control process. With NotifyAccess= equal to "exec" or "all" this process is allowed to signal startup completion. However, when it sends READY=1, the reloading is completed and we go straight into SERVICE_RUNNING anyway. This is more something of a bug than relevant to the current discussion.

Reply:
But if you do use PAM in such a case, then READY=1 will not be accepted from the process, because it moved to a different cgroup in that case.

or to say this differently: as soon as your process that is supposed to send READY=1 joins a different cgroup, then all bets are off! I don't think it's really unexpected that if you set NotifyAccess=all, but then migrate your process out of the cgroup, and do not make it main or control PID, that the notification messages won't reach the right unit. I mean, in that case you did the most you can do, to *not* be associated with the original unit anymore, hence of course the messages won't be accepted.

We should probably document near PAMName= that Type=notify and PAMName= doesn't really mix well.